nl/seg6_linux.go

```diff
@@ -0,0 +1,121 @@
+package nl
+
+import (
+	"errors"
+	"fmt"
+	"net"
+)
+
+type Ipv6SrHdr struct {
+	nexthdr       uint8
+	hdrlen        uint8
+	typ           uint8
+	segments_left uint8
+	first_segment uint8
+	flags         uint8
+	reserved      uint16
+
+	Segments []net.IP
+}
+
+func (s1 *Ipv6SrHdr) Equal(s2 Ipv6SrHdr) bool {
+	if len(s1.Segments) != len(s2.Segments) {
+		return false
+	}
+	if len(s1.Segments) != 0 {
+		for i := range s1.Segments {
+			if s1.Segments[i].Equal(s2.Segments[i]) != true {
+				return false
+			}
+		}
+	}
+	return s1.nexthdr == s2.nexthdr &&
+		s1.hdrlen == s2.hdrlen &&
+		s1.typ == s2.typ &&
+		s1.segments_left == s2.segments_left &&
+		s1.first_segment == s2.first_segment &&
+		s1.flags == s2.flags
+	// reserved doesn't need to be identical.
+}
+
+// seg6 encap mode
+const (
+	SEG6_IPTUN_MODE_INLINE = iota
+	SEG6_IPTUN_MODE_ENCAP
+)
+
+// number of nested RTATTR
+// from include/uapi/linux/seg6_iptunnel.h
+const (
+	SEG6_IPTUNNEL_UNSPEC = iota
+	SEG6_IPTUNNEL_SRH
+	__SEG6_IPTUNNEL_MAX
+)
+const (
+	SEG6_IPTUNNEL_MAX = __SEG6_IPTUNNEL_MAX - 1
+)
+
+func EncodeSEG6Encap(mode int, srh Ipv6SrHdr) ([]byte, error) {
+	nsegs := len(srh.Segments) // nsegs: number of segments
+	if nsegs == 0 {
+		return nil, errors.New("EncodeSEG6Encap: No Segment in srh")
+	}
+	b := make([]byte, 12, 12+len(srh.Segments)*16)
+	native := NativeEndian()
+	native.PutUint32(b, uint32(mode))
+	srh.nexthdr = 0                     // use 0 when calling netlink
+	srh.hdrlen = uint8(16 * nsegs >> 3) // in 8-octets unit
+	srh.typ = 4                         // Routing Type to be assigned by IANA (suggested value: 4)
```

Comment:
Shouldn't we have a constant for this as does the kernel code 
```
#define IPV6_SRCRT_TYPE_4	4	/* Segment Routing with IPv6 */
```

Reply:
Defined them in nl/syscall.go
```
// routing header types
const (
    IPV6_SRCRT_STRICT = 0x01 // Deprecated; will be removed
    IPV6_SRCRT_TYPE_0 = 0    // Deprecated; will be removed
    IPV6_SRCRT_TYPE_2 = 2    // IPv6 type 2 Routing Header
    IPV6_SRCRT_TYPE_4 = 4    // Segment Routing with IPv6
)
```
And changed to: `srh.routingType = IPV6_SRCRT_TYPE_4 // Routing Type assigned by IANA`